### DISCUSSION TITLE
Environment variable doesn't work

### DISCUSSION THREAD

**Discussion by Frank-Schaeske** [neuefische/questions#652] [Web] [answered]
Hi,
in my Form component I'm fetching data from an external API (RapidAPI) which works fine as long as the API-Key is hard coded. When I try to replace the key with an environment variable the fetch doesn't work any longer. 

I have committed and pushed my current version. [You can find it here.](https://github.com/Frank-Schaeske/ScoutingHelper/blob/feature/fetch-api/components/Form/index.js) The branch name is `feature/fetch-api`.

My .env.local file is in the root directory and includes one line of code:
```RAPID_API_KEY=9037d3f3c3msh61cd09ee7270a1dp18a502jsn2f04d269aXXX```
(In reality XXX is different) 

Description of the error:
How to reproduce the error?
When fetching with the environment variable instead of the hard coded key the error occurs. 
Where and when does it occur?
Whenever I try to fetch with the variable.
What is expected, what happens?
I expect the fetch to run exactly as it does with the hard coded key.
What are the error messages?
Instead of the desired object with the data I get the following object: ```{message: 'You are not subscribed to this API.'}```.
What have I already tried:
I have read through the [documentation](https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables) and asked ChatGPT.

Many thanks for your help!

**Comment by ekat-br** [comment]
Hi Frank, 

this is the first time I have dealt with this issue, but I am trying my best to help you. 

I skimmed your code so far and tried the following in the .env.local: `NEXT_PUBLIC_RAPID_API_KEY=` and `const rapidApiKey = process.env.NEXT_PUBLIC_RAPID_API_KEY` instead of `RAPID_API_KEY` and it worked.

Now the explanation why it works (with the help of chatgpt) : 

- Without the "NEXT_PUBLIC_" prefix, the environment variable is treated as a server-side variable, meaning it is only accessible on the server-side of your application. This is a security measure to prevent sensitive information like API keys from being exposed to the client-side (browser) where users can potentially view the source code.
- In your code, you are using the API key in a client-side context to make a fetch request to an external API (RapidAPI). Since the API key is a sensitive piece of information, Next.js by default restricts its accessibility on the client-side to prevent accidental exposure.
- When you use the "NEXT_PUBLIC_" prefix for the environment variable (e.g., "NEXT_PUBLIC_RAPID_API_KEY"), you are explicitly telling Next.js that this variable is intended to be accessed on the client-side as well. Next.js treats such variables differently, and their values are safely injected into the client-side JavaScript code during the build process.
- During the build process, Next.js takes the values of the "NEXT_PUBLIC_" environment variables and injects them as static values into the client-side code. This means the actual sensitive values (e.g., API keys) are not exposed at runtime when users access your website in their browsers.

I hope this helps you. It would probably be good if others would also look in and correct me if I am wrong with my explanation. 🙏

**Reply by Frank-Schaeske** [reply]
Hi Katja, 

thank you very much for your effort! When I asked ChatGPT about my problem I got the following remark: 

Keep in mind that you should not include sensitive information in public environment variables, as they can be accessed by anyone with access to your client-side JavaScript code. For sensitive information like API keys, it's better to use regular environment variables (without the NEXT_PUBLIC_ prefix) that will only be available on the server-side.

Could somebody clarify, if it is OK/safe to work with the NEXT_PUBLIC_ prefix in my case?

Thanks in advance!

**Comment by Roland-Hufnagel** [comment, marked as answer]
Hello Frank, 

if you work with sensitive data like your API_KEY, you should not fetch from the frontend (your component) but from the backend. This is the place where you can access your process.env.local. What I mean:
Create an API route (pages/api/player/index.js) and write that handler function here. This is the 'backend'. Here you can access your .env.local and the sensitive data is not available in the frontend. 
In the frontend then fetch your own API-route (that makes the rapid-fetch and returns the data). 
This should work.
